packetbeat/protos/drda/drda.go

```diff
@@ -0,0 +1,572 @@
+package drda
+
+import (
+	"fmt"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"strings"
+	"time"
+
+	"github.com/Intermernet/ebcdic"
+	"github.com/elastic/beats/packetbeat/procs"
+	"github.com/elastic/beats/packetbeat/protos"
+	"github.com/elastic/beats/packetbeat/protos/tcp"
+	"github.com/elastic/beats/packetbeat/publish"
+	"github.com/urso/ucfg"
+)
+
+/*
+
+Parse DRDA Protocol used by IBM DB2, Informix or Apache Derby (and possibly other databases)
+
+Limitations:
+- Raw message not supported (guess that makes not much sense here)
+- No unit tests yet
+- Too few system tests
+- Check that all codepoints are covered (also "esoteric" like those from IMS)
+
+Dependencies:
+- https://github.com/Intermernet/ebcdic
+
+*/
+
+type parseState int
+
+const (
+	drdaStateDDM parseState = iota
+	drdaStateParameter
+)
+
+var stateStrings []string = []string{
+	"DDM",
+	"Parameter",
+}
+
+func drdaAbbrev(codepoint uint16) string {
+	abbrev := drda_abbrev[codepoint]
+
+	if abbrev == "" {
+		return fmt.Sprint("unknown_", codepoint)
+	}
+
+	return abbrev
+}
+
+func (state parseState) String() string {
+	return stateStrings[state]
+}
+
+func (drda *Drda) getTransaction(k common.HashableTcpTuple) *DrdaTransaction {
```

Comment:
Tips: some modules have been updated to not store transactions globally:
- but within the TCP transaction. Simplifies transaction lookup + connection timeout is handled in TCP module.
- store message in list instead of creating intermediate transaction object (helps with pipelining in DB clients using async IO)

I recently create a cookiecutter module generating loads of boilerplate code for transaction handling (mostly parser needs to be implemented): https://github.com/urso/packetbeat-tcp-generator